ratpack-test/src/main/java/ratpack/test/http/proxy/TestHttpProxyServer.java

```diff
@@ -0,0 +1,160 @@
+package ratpack.test.http.proxy;
+
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.handler.codec.http.HttpServerCodec;
+import io.netty.handler.logging.LogLevel;
+import io.netty.handler.logging.LoggingHandler;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import ratpack.http.client.ProxyCredentials;
+import ratpack.http.client.internal.DefaultProxyCredentials;
+import ratpack.test.http.proxy.internal.Authority;
+import ratpack.test.http.proxy.internal.ProxyAuthenticationHandler;
+import ratpack.test.http.proxy.internal.ProxyClientHandler;
+import ratpack.util.Exceptions;
+
+import javax.net.ssl.SSLException;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static ratpack.test.http.proxy.internal.ProxyHandlerNames.HTTP_CODEC_HANDLER;
+
+// TODO Add class-level and method-level Javadoc
+public class TestHttpProxyServer implements AutoCloseable {
```

Comment:
There wasn't a suitable embeddable, Java-based test HTTP proxy available from the open source community, so I decided to write one (Netty has one in their internal test code, but Netty doesn't distribute their internal test code).

The Ratpack HttpClient (and Netty) currently only supports using the CONNECT method, so the TestProxyServer only supports CONNECT as well.

TestHttpServer is written using just Netty (and not Ratpack). The CONNECT method of connecting to an HTTP Proxy requires establishing a low-level TCP tunnel, which is hard to accomplish with Ratpack.

I haven't yet gotten a chance to write the Javadoc for this class, and how it is configured is a bit messy right now.

I added this to Ratpack's public test library so that applications built on Ratpack can use it.

Reply:
No problem now that it's internal.